segway/input_master.py

```diff
@@ -139,201 +140,253 @@ def make_zero_diagonal_table(length):
-
-        # fudge the minimum by a very small amount. this is not
-        # continuous, but hopefully we won't get values where it
-        # matters
-        # XXX: restore this after GMTK issues fixed
-        # if min_track == 0.0:
-        #     min_track_fudged = FUDGE_TINY
-        # else:
-        #     min_track_fudged = min_track - ldexp(abs(min_track), FUDGE_EP)
-
-        # this happens for really big numbers or really small
-        # numbers; you only have 7 orders of magnitude to play
-        # with on a float32
-        min_track_f32 = float32(min_track)
-
-        assert min_track_f32 - float32(ABSOLUTE_FUDGE) != min_track_f32
-        return min_track - ABSOLUTE_FUDGE
-
-    def make_segnames(self):
-        return format_indexed_strs("seg", self.num_segs)
-
-    def make_subsegnames(self):
-        return format_indexed_strs("subseg", self.num_subsegs)
-
-    def get_template_component_suffix(self, component_number):
-        """Returns the subsitution for the component suffix in the GMTK model
-        template. Empty if there is only one component"""
-        if self.num_mix_components == 1:
-            return ""
-        else:
-            return "_component{}".format(component_number)
-
-    def generate_tmpl_mappings(self):
-        # need segnames because in the tied covariance case, the
-        # segnames are replaced by "any" (see .make_covar_spec()),
-        # and only one mapping is produced
-        num_subsegs = self.num_subsegs
-
-        track_groups = self.track_groups
-        num_track_groups = self.num_track_groups
-
-        for seg_index, segname in enumerate(self.make_segnames()):
-            seg_offset = num_track_groups * num_subsegs * seg_index
-
-            for subseg_index, subsegname in enumerate(self.make_subsegnames()):
-                subseg_offset = seg_offset + (num_track_groups * subseg_index)
-
-                for track_group_index, track_group in enumerate(track_groups):
-                    track_offset = subseg_offset + track_group_index
-                    head_trackname = track_group[0].name
-
-                    # XXX: change name of index to track_offset in templates
-                    # XXX: change name of track_index to track_group_index
-                    yield dict(seg=segname, subseg=subsegname,
-                               track=head_trackname, seg_index=seg_index,
-                               subseg_index=subseg_index,
-                               track_index=track_group_index,
-                               index=track_offset,
-                               distribution=self.distribution)
+    def __str__(self):
+        return make_spec(self.type_name, self.generate_objects())
 
     def make_data(self):
         """
         override this in subclasses
-
         returns: container indexed by (seg_index, subseg_index, track_index)
         """
         return None
 
-    def generate_objects(self):
+    def get_head_track_names(self):
```

Comment:
Is there a difference between this and `self.track_names` ?

Reply:
Yes, in the simpleconcat case: 
```
segway "$cluster_arg" \
    train --include-coords="../include-coords.bed" --num-labels=4 \
    --track=testtrack1A,testtrack1B --track=testtrack2A,testtrack2B \
    "../simpleconcat.genomedata" traindir
```
the `get_head_track_names` would return ["testtrack1A", "testtrack2A"], but `self.track_names` would be ["testtrack1A", "testtrack1B", "testtrack2A", "testtrack2B"]